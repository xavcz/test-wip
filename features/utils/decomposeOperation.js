import { parse, print } from 'graphql';

// from a GraphQL AST or a GraphQL string, get both
// source: string
// document info extracted from the AST
export default operation => {
  const isString = typeof operation === 'string';

  return {
    source: isString ? operation : print(operation),
    information: extractDocumentInfo(isString ? parse(operation) : operation),
  };
};

// note: should be extracted elsewhere
// extract operation name and first selection name
// input: parsed graphql document (gql`` / graphql file)
// gql`query yoUoYo {
//   getBalance { ... }
// }`
// output: { operation: 'query', operationName: 'yoUoYo', selectionName: 'getBalance' }
const extractDocumentInfo = document => {
  const { definitions } = document;
  const [firstDef] = definitions;

  const {
    selectionSet: { selections },
    operation,
    name: operationNameObject,
  } = firstDef;

  if (!operationNameObject) {
    if (operation === 'query') {
      throw new Error(
        'Implicit query are not allowed! Please specify an operation name.'
      );
    }

    throw new Error(
      'Unnamed operations are not allowed! Please specify an operation name.'
    );
  }

  if (selections.length > 1) {
    throw new Error(
      'More than 1 selection is defined in this operation! Please specify only 1 selection'
    );
  }

  const [firstSelection] = selections;
  const { name: { value: selectionName } } = firstSelection;

  return { operation, operationName: operationNameObject.value, selectionName };
};
