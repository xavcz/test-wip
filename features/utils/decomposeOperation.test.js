import gql from 'graphql-tag';
import decomposeOperation from './decomposeOperation';

describe('decomposeOperation', () => {
  it('should return a GraphQL source & infos about it from a GraphQL source', () => {
    const operationAsSource = 'query hello { world }';

    expect(decomposeOperation(operationAsSource)).toMatchSnapshot();
  });

  it('should return a GraphQL source & infos about it from a GraphQL source', () => {
    const operationAsDocument = gql`
      query hello {
        world
      }
    `;

    expect(decomposeOperation(operationAsDocument)).toMatchSnapshot();
  });

  it('should fail if the operation is an implicit query', () => {
    const implicitQuery = gql`
      {
        implicitQuery
      }
    `;

    expect(() =>
      decomposeOperation(implicitQuery)
    ).toThrowErrorMatchingSnapshot();
  });

  it('should fail if the operation is an unnamed operation', () => {
    const disallowedDoc = gql`
      mutation {
        noOperationName
      }
    `;

    expect(() =>
      decomposeOperation(disallowedDoc)
    ).toThrowErrorMatchingSnapshot();
  });

  it('should fail if the operation has more than 1 selection', () => {
    const disallowedDoc = gql`
      mutation tooComplexForThisProgram {
        firstSelection {
          abc
        }
        secondSelection {
          xyz
        }
      }
    `;

    expect(() =>
      decomposeOperation(disallowedDoc)
    ).toThrowErrorMatchingSnapshot();
  });
});
