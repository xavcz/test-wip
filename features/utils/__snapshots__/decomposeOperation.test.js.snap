// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decomposeOperation should fail if the operation has more than 1 selection 1`] = `"More than 1 selection is defined in this operation! Please specify only 1 selection"`;

exports[`decomposeOperation should fail if the operation is an implicit query 1`] = `"Implicit query are not allowed! Please specify an operation name."`;

exports[`decomposeOperation should fail if the operation is an unnamed operation 1`] = `"Unnamed operations are not allowed! Please specify an operation name."`;

exports[`decomposeOperation should return a GraphQL source & infos about it from a GraphQL source 1`] = `
Object {
  "information": Object {
    "operation": "query",
    "operationName": "hello",
    "selectionName": "world",
  },
  "source": "query hello { world }",
}
`;

exports[`decomposeOperation should return a GraphQL source & infos about it from a GraphQL source 2`] = `
Object {
  "information": Object {
    "operation": "query",
    "operationName": "hello",
    "selectionName": "world",
  },
  "source": "query hello {
  world
}
",
}
`;
